contao/core-bundle
# [RFC] Symfony RememberMe Service Implementation

This is the PR for Issue #454 

The current status is completely working but I need Feedback before going further
## What is this PR for:

Providing autologin functionality across different devices for a frontend user. The current implementation only allows autologin for one device.
## Questions
1. **Make the Implementation optional, parallel or replace?** We could add _Allow autologin from multiple devices_ to tl_settings. See https://github.com/contao/core-bundle/issues/454#issuecomment-217706318 for infos on it.  
   I would vote for optional in combination with parallel for the most user comfort with deprecation warning for Contao 5. (Switching from normal autologin to multi-autologin would not break autologin cookies)
2. **How to handle the _cookie lifetime_ param?** Since the RememberMeService needs this setting at startup and cannot be changed later I had to put it in the DI Container for now. There the `$GLOBALS['TL_CONFIG']['autologin']` is not available atm. So we either had to 
   - remove it from tl_settings
   - write it back to the parameters.yml (bad practice?) 
   - load the localconfig.php with default.php (as fallback) in DI / CoreExtension (also bad practice?)
   - write it to another, dynamic configuration.yml and load in  DI / CoreExtension
## Further Implementation Infos
- Currently the rememberMeService lives beside the current autologin implementation (which is no problem, they do not harm each other in any way)
- The contao.logger you see in the services.yml / logger class is for development. The Symfony rememberMe Service class can _only_ handle a psr-3 compatible logger, so I made a quick wrapper for it. Will be removed 
- I had to extend the Contao UserProvider to make it compatible to the remember service. The current implementation does not allow to really load a user. I'm sure this was done on purpose by @aschempp but don't know why
- The rememberMeService requires The User Object to implement the `UserInterface`. I moved that part to the `User` class instead of the `FrontendUser`, so this has not to be changed for future auth implementations where the backend is also involved.
- The authenticate method diff looks bigger than it is. I had to move one part and restructure a little bit to avoid code duplication
- Code is not very pretty atm. Are there any Code / Style Conventions? Didn't found any info on it
## How To test
- Setup a Contao Installation with a login form and autologin enabled
- Create a member user
- Login to the frontend
- Delete the FE_AUTH and FE_AUTOLOGIN cookie, reload the site (currently the rememberMeService uses it's own Cookie FE_AUTOLOGIN_MULTI). See the user should still be logged in. 
## Todos
- [ ] Write Unittests
- [x] Implementation of Question 1,2

---

~~I had a to rush a little bit when I was half through the PR, because guests arrived for BQ, but I wanted to get out to you. Will reread that later and _optimize_~~ <- done :smile:
